includes/class-wcpdf-admin.php

```diff
@@ -1039,7 +1044,7 @@ public function ajax_crud_document() {
 
 		try {
 			// Allow getting documents even if their prerequisite documents doesn't exist.
-			add_filter( 'wpo_wcpdf_prerequisite_documents', '__return_empty_array()' );
+			add_filter( 'wpo_wcpdf_document_do_prerequisites_exist', '__return_true' );
```

Comment:
Maybe something like this would work below line `1049`? Because we are not forcing the document generation at this stage, so it's safe to call the document first.

```php
if ( false === apply_filters( 'wpo_wcpdf_bypass_prerequisites_restriction', true, 'ajax_crud_document', $document ) && ! empty( $document->get_prerequisite_documents() ) ) {
	return;
}
```

This can be also applied to the `data_input_box_content()` inside the loop I think.

And maybe inside `is_allowed()`?

Reply:
We can add checking prerequisites inside the `is_allowed()` method, inside the condition you mentioned, but again we need to change the value that the `wpo_wcpdf_bypass_prerequisites_restriction` filter returns, before line `1049`, since otherwise `wcpdf_get_document()` method won't return the document.
So, I think it's again adding a filter above the line `1049`.

Reply:
In this commit: https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/pull/757/commits/08cff804ddb50a6a542f62f868a7a76c5fc59680, I added again checking prerequisites to `is_allowed()` method, but made it `false` as default, since otherwise it'll allow creation of documents without prerequisites, and also used the filter you mentioned.
Please let me know what you think.

Reply:
You could avoid having it inside `is_allowed()`, and call it when you need to check the `is_allowed()` output. I think returning an empty array seems strange to me and a bit like hacking.

Reply:
If we don't add checking prerequisites to the `is_allowed()` method, the `wcpdf_get_document()` method will return a document even if the prerequisites of it don't exist. so I thought maybe it should be inside that as default?

> I think returning an empty array seems strange to me and a bit like hacking.

You're right.
In the recent commit, I used the filter you mentioned: `wpo_wcpdf_bypass_prerequisites_restriction`. What do you think about it?

Reply:
Yes, without sending the last parameter (`$init`) to the `wcpdf_get_document()` function, it won't generate it, but it'll still return the document. For example, a customer can manipulate the URL to get a receipt document. This is the URL to get the invoice:
`.../wp-admin/admin-ajax.php?action=generate_wpo_wcpdf&document_type=invoice&order_ids=1235...`
So they can change the `document_type` parameter and get the receipt. While the receipt is not actually saved, they can get a PDF. 
If it's OK, then I can change that filter value to `true` as default and add the condition you mentioned after the line `1049`.

Reply:
I understood your point.
So we add this condition to `data_input_box_content()`, `generate_document_ajax()`, and `is_allowed()` and make the filter value `true` by default, but for `generate_document_ajax()`, we make it `false` by default.

In this case, I like to propose adding the `do_prerequisites_exist()` method again, but with small changes:

```php
public function do_prerequisites_exist( string $context = '', bool $default = true ): bool {
	foreach ( $this->get_prerequisite_documents() as $type ) {
		$document = wcpdf_get_document( $type, $this->order );

		if ( ! $document || ! $document->exists() ) {
			return apply_filters( 'wpo_wcpdf_bypass_prerequisites_restriction', $default, $context, $this );
		}
	}

	return true;
}
```

What do you think?

Reply:
In this case we don't need the filter do we? I think you might need to check if `get_prerequisite_documents()` is not empty?

Reply:
We can still keep the filter; it will allow us to bypass the `false` value, which I think would be good.

Another issue I see (with both approaches; using the condition you mentioned or the `do_prerequisites_exist()` function) is that if a customer changes the `document_type` in the URL, as I mentioned, we can check here to prevent displaying the document:

https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/blob/main/includes/class-wcpdf-main.php#L545-L547

Like this:
```php
if ( $document && $document->do_prerequisites_exist( 'generate_document_ajax', false ) ) {
```

It'll prevent the document from being displayed but not prevent the document from being generated. So, the receipt will be generated but not displayed.

Reply:
This issue will happen everywhere when the third parameter of `wcpdf_get_document()` is `true`.

One way I can think of is to use the filter provided in the `do_prerequisites_exist()` method to change the `$bypass` value and also make the `$bypass` default value `false`. It'll be still the same as the previous solution; using a filter, but maybe seems less hacking? Since we don't change the prerequisites value.

Reply:
I think you need to check the prerequisite before forcing the document generation here: https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/blob/main/includes/class-wcpdf-main.php#L545

Reply:
Since a document object is required to call the `do_prerequisites_exist()`, to check the prerequisite before forcing the document generation, I must call the `wcpdf_get_document()` again without the third parameter. I'm not sure if this is a good approach.

Reply:
You can get the document from here: `$document = WPO_WCPDF()->documents->get_document( $document_type, $order );`

Reply:
I added that to `is_allowed()` method like this:
```php
if ( ! $this->do_prerequisites_exist( $context ) ) {
	return false;
}
```

But since it'll bypass by default, it won't affect it. If we change the bypass to `false` by default, then we need to use the filter to change it anywhere we want, for example, for metabox.

I think checking prerequisites is a requirement for the `wcpdf_get_document()` method, so maybe it's better to keep the bypass `false` as default?

Also, we can use the `WPO_WCPDF()->documents->get_document()` method here, but I think it'll make the usage of `wcpdf_get_document()` and it should be something that should be considered for using of `wcpdf_get_document()` method.

If you agree, I will revert the related change to this matter and open another PR to address this issue, which is related to issue #760.

Reply:
> If you agree, I will revert the related change to this matter and open another PR to address this issue, which is related to issue https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/issues/760.

Yes, it's better, we continue from there.